### ISSUE TITLE
Inconsistent prescribed iteration limits with stencil offsets

### ISSUE THREAD

**Issue by mlange05** [opesci/devito#425]
Assume the following setup code:
```
grid = Grid(shape=(11, 11))
x, y = grid.dimensions
a = Function(name='a', grid=grid)
a.data[:] = 1.
A = a.indexed
```

Now, if I run an operator with no stencil offsets, I get what I'd expect:
```
Operator(Eq(a, a + a))(x_s=3, x_e=7)
assert (a.data[:3, :] == 1.).all()
assert (a.data[3:7, :] > 1.).all()
assert (a.data[7:, :] == 1.).all()
```
If I add an operator with a stencil offset in `x` though, I get a positive shift (`+1`) in iteration space of _both_, `x_s` and `x_e`:
```
a.data[:] = 1.
Operator(Eq(a, a + (A[x-1, y] + A[x+1, y]) / 2.))(x_s=3, x_e=7)
print(a.data)
```
yields
```
[[ 1.     1.     1.     1.     1.     1.     1.     1.     1.     1.     1.   ]
 [ 1.     1.     1.     1.     1.     1.     1.     1.     1.     1.     1.   ]
 [ 1.     1.     1.     1.     1.     1.     1.     1.     1.     1.     1.   ]
 [ 1.     1.     1.     1.     1.     1.     1.     1.     1.     1.     1.   ]
 [ 2.     2.     2.     2.     2.     2.     2.     2.     2.     2.     2.   ]
 [ 2.5    2.5    2.5    2.5    2.5    2.5    2.5    2.5    2.5    2.5    2.5  ]
 [ 2.75   2.75   2.75   2.75   2.75   2.75   2.75   2.75   2.75   2.75 2.75 ]
 [ 2.875  2.875  2.875  2.875  2.875  2.875  2.875  2.875  2.875  2.875 2.875]
 [ 1.     1.     1.     1.     1.     1.     1.     1.     1.     1.     1.   ]
 [ 1.     1.     1.     1.     1.     1.     1.     1.     1.     1.     1.   ]
 [ 1.     1.     1.     1.     1.     1.     1.     1.     1.     1.     1.   ]]
```
This final result is shifted in `x`, which seems non-sensical. I would either expect the iteration space to be the exact same as before (`x[3:7]`) or to contract due to the stencil offset, ie. `x[4:6]`, but I see no justification for a shifted iteration space `x[4:8]`. 

@navjotk What is the current convention, and what iteration space should I expect. Is this simply a bug or a design feature?

**Comment by navjotk** [comment]
As far as I understand, this follows the specification you gave me in https://github.com/opesci/devito/pull/345#issuecomment-336134755 . 
What is different here compared to what you described in the comment I linked above?

**Comment by mlange05** [comment]
This is why context is so important: the comment clearly states that this is for the _upcoming switch to properly using the concept of a "domain", as termed in YASK._ This change has not yet happened, as far as I'm aware, leaving the current behaviour out of sync with the rest of the API.

**Comment by navjotk** [comment]
So what was supposed to be the pre-switch behaviour according to that thread?

**Comment by mlange05** [comment]
That's the question I was asking you: direct translation (`[3:7] -> [3:7]`) or contraction (`[3:7] -> [4:6]`)? Either can be justified, at least to my mind, while the implicit shift (`[3:7] -> [4:8]`) has no real justification as far as I can see.

Whatever we decide, the start/end + stencil offsets case needs testing!

**Comment by mloubout** [comment]
@fablup, fixed by domain switch I think.

**Comment by FabioLuporini** [comment]
fixed by #507
